meilisearch-core/src/raw_indexer.rs

```diff
@@ -113,52 +141,57 @@ impl<A: AsRef<[u8]>> RawIndexer<A> {
     }
 }
 
-fn index_token<A>(
+fn process_tokens<'a>(tokens: impl Iterator<Item = Token<'a>>) -> impl Iterator<Item = (usize, Token<'a>)> {
+    tokens
+        .scan((0, None), |(offset, sepkind), token| {
+                match token.kind {
+                    TokenKind::Word | TokenKind::StopWord | TokenKind::Any => {
+                        *offset += match *sepkind {
+                            Some(SeparatorKind::Hard) => 8,
+                            Some(SeparatorKind::Soft) => 1,
+                            None => 0,
+                        };
+                        *sepkind = None;
+                    }
+                    TokenKind::Separator(SeparatorKind::Hard) => {
+                        *sepkind = Some(SeparatorKind::Hard);
+                    }
+                    TokenKind::Separator(SeparatorKind::Soft) if sepkind.is_none() => {
+                        *sepkind = Some(SeparatorKind::Soft);
+                    }
+                    _ => (),
+                }
+            Some((*offset, token))
+        })
+    .filter(|(_, t)| t.is_word())
+}
```

Comment:
```suggestion
fn process_tokens<'a>(tokens: impl Iterator<Item = Token<'a>>) -> impl Iterator<Item = (usize, Token<'a>)> {
    tokens
        .scan((0, false), |(offset, hard_separated), token| {
                match token.kind {
                    TokenKind::Word | TokenKind::StopWord | TokenKind::Any => {
                        *offset += match *hard_separated {
                            true => 8,
                            false => 1,
                        };
                        *hard_separated = false;
                    }
                    TokenKind::Separator(SeparatorKind::Hard) => {
                        *hard_separated = true;
                    }
                    _ => (),
                }
            Some((*offset, token))
        })
    .filter(|(_, t)| t.is_word())
}
```

In the case of Chinese script, we would have a lot of words in the same position because they are not separated with separators, and so, the offset would never be incremented.

Reply:
This will not work... but I have an idea :)